mmv1/third_party/terraform/.teamcity/components/inputs/diff_check.go

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+/*
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+// This file is controlled by MMv1, any changes made here will be overwritten
+
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"regexp"
+)
+
+func main() {
+	repo := os.Args
+	services := fmt.Sprintf("../../../%v/services/...", repo[1])
+	fmt.Println(services)
+	cmd := exec.Command("go", "list", services)
+	stdout, err := cmd.Output()
+	if err != nil {
+		fmt.Println(err.Error())
+		return
+	}
+
+	pattern := regexp.MustCompile(`github\.com\/hashicorp\/terraform-provider-(google|google-beta)\/(google|google-beta)\/services\/(?P<service>\w+)`)
+
+	// Template to convert "key: value" to "key=value" by
+	// referencing the values captured by the regex pattern.
+	template := []byte("$service\n")
+
+	googleServices := []byte{}
```

Comment:
Instead of collecting this information as a long slice of bytes, I'd recommend collecting a list of the service names as human readable strings. This would make the tool easier to debug during development and also help with making actionable error messages when a service is identified as missing.

The loop over submatches on lines 42-46 could contain something like:

```
dst := []byte{}
service := pattern.Expand(dst, template, stdout, submatches)
googleServices = append(googleServices, string(service))
```

---

Following that, another idea to think about might be storing the found services in a map where the key is the service name, instead of using a list. This would make it much easier to identify if a service is present or not:

```
// assuming data gets stored as map[string]bool (value in key-value pair not super important)

// from processing services_ga.kt
servicesInTeamCity := map[string]bool {
  "serviceA" : true,
  // serviceB isn't a key because it's missing from the services_ga.kt file used to make the map
  "serviceC" : true,
  ...
}

// from processing stdout of go list command
servicesInGaProvider := map[string]bool {
  "serviceA" : true,
  "serviceB" : true,
  "serviceC" : true,
  ...
}

for service, _ := range servicesInGaProvider {
  val, ok := servicesInTeamCity[service]; !ok {
    fmt.Printf("service %s is in the GA provider but not in services_gs.kt!", service)
  }
}
```